File: client/src/test/scala/stasis/test/specs/unit/client/mocks/MockChecksum.scala
package stasis.test.specs.unit.client.mocks

import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

import scala.concurrent.Future

import org.apache.pekko.stream.Materializer

import stasis.client.analysis.Checksum
import stasis.test.specs.unit.client.mocks.MockChecksum.Statistic

class MockChecksum(checksums: Map[Path, BigInt]) extends Checksum {
  private val stats: Map[Statistic, AtomicInteger] = Map(
    Statistic.ChecksumCalculated -> new AtomicInteger(0)
  )

  override def calculate(file: Path)(implicit mat: Materializer): Future[BigInt] = {
    stats(Statistic.ChecksumCalculated).incrementAndGet()
    checksums.get(file) match {
      case Some(checksum) =>
        Future.successful(checksum)

      case None =>
        Future.failed(new IllegalArgumentException(s"No checksum found for file [$file]"))
    }
  }

  def statistics: Map[Statistic, Int] = stats.view.mapValues(_.get()).toMap
}

object MockChecksum {
  sealed trait Statistic
  object Statistic {
    case object ChecksumCalculated extends Statistic
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockChecksum.kt
package stasis.test.client_android.lib.mocks

import stasis.client_android.lib.analysis.Checksum
import java.math.BigInteger
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

class MockChecksum(val checksums: Map<Path, BigInteger>) : Checksum {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.ChecksumCalculated to AtomicInteger(0),
    )

    override suspend fun calculate(file: Path): BigInteger {
        stats[Statistic.ChecksumCalculated]?.getAndIncrement()

        when (val checksum = checksums[file]) {
            null -> throw IllegalArgumentException("No checksum found for file [$file]")
            else -> return checksum
        }
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object ChecksumCalculated : Statistic()
    }
}
